File: LeetCode/validate_binary_search_tree.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return isValidBST(root, LONG_MAX, LONG_MIN);
    }
private:
    bool isValidBST(TreeNode* node, long max, long min){
        if(!node){
            return true;
        }else if(node->val >= max || node->val <= min){
            return false;
        }
        return isValidBST(node->left, node->val, min) && isValidBST(node->right, max, node->val);
    }
};

int main(void){
    TreeNode root(2);
    TreeNode left(1);
    TreeNode right(3);
    root.left = &left;
    root.right = &right;
    Solution sol;
    cout << sol.isValidBST(&root) << endl;
    return 0;
}

File: LeetCode/validate_binary_search_tree.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def isValidBST(self, root):
        return self.isValid(root, -float('inf'), float('inf'))
    
    def isValid(self, node, min_val, max_val):
        if not node:
            return True
        
        if node.val <= min_val or node.val >= max_val:
            return False
        
        return self.isValid(node.left, min_val, node.val) and\
            self.isValid(node.right, node.val, max_val)

if __name__ == "__main__":
    root = TreeNode(2)
    root.left = TreeNode(1)
    root.right = TreeNode(3)
    sol = Solution()
    res = sol.isValidBST(root)
    print(res)